### ISSUE TITLE
Charged particles

### ISSUE THREAD

**Issue by Ceres90** [Syrocco/Graphical-EDMD#2]
Do you think this code could be modifed to include particles with different properties like charge, ie positive particles would stick if they touched negative particles , but repel or not stick to other positive particles?

**Comment by Syrocco** [comment]
Yes surely. But the interaction would need to be discontinuous. 
In EDMD.h, you can had a property to the type particle that would be the charge. Then, you could use the already implemented algorithm for square shoulder  (U>0)   and square well potential (U<0) to behave differently according to the type of the two particles colliding. If both particles have the same charge, you’d want to use the interaction with U > 0, otherwise, you’d use the interaction with U < 0.

**Comment by Syrocco** [comment]
I added a minimal charged particle example. You can look at the (before) last commit change to see what I did. You can also compile the new code. You have a new button:
![image](https://github.com/Syrocco/Graphical-EDMD/assets/37348774/64d549a1-d137-40f5-a8c9-9b627e5d0e8c) alongside the % of charged particle (with the same charge, there're only + and - charges, with same absolute value of charge). You can control the attraction/repulsion strenght between different/same charges using U (-U and U are the same, only the absolute value of U matters). You can also display different charge in different color with the dropdown:
![image](https://github.com/Syrocco/Graphical-EDMD/assets/37348774/404ee72b-3643-49d3-bd33-5639629229f1)

This will create crystal like structures of alternating charge: 
![image](https://github.com/Syrocco/Graphical-EDMD/assets/37348774/c58d3a7b-9061-4ffe-b0e9-b68bd9d71c11)

**Comment by Syrocco** [comment]
That was a cool idea. At large interaction distance, you can obtain very nice patterns:
![image](https://github.com/Syrocco/Graphical-EDMD/assets/37348774/ad93ff97-bb07-429a-915b-c5ef8a7ddac0)

In physics, this is called Frustration. You have geometrical constraint (here the volume and the number of particle are fixed) and the system tries to minimize its energy. Off course, every particles cannot have only the neighbors it would like to have, so it has to do some compromise. On top of that, these kind of systems are known to have a very slow dynamics (we call them Glasses) and do not manage to find the minimal energy state easily, so for a very very very long time, you only see a system that is in a local minimum of energy and not a global one.

**Comment by Ceres90** [comment]
wow thanks for the quick response and change in code!!!  this program could have really interesting biology applications with just a couple more additions .  I work in a lab where we study nanoparticle interactions with various pathogens (viruses and bacteria) and we use zeta potential (which is the measure of the particles charge ) to predict interactions but we currently only perform wetlab experiements  I saw your comment on youtube , what makes the neutral particle more difficult(just curious not trying to be a jerk) ?  Would it be possible to have certain particles that follow the old interaction equation that didn;t include the charge rule? (i have no coding experience so maybe thats a dumb suggestion)

**Comment by Ceres90** [comment]
<img width="325" alt="idea" src="https://github.com/Syrocco/Graphical-EDMD/assets/95306501/cc094967-299a-4c90-9db7-9c71709bbc7d">
I attached a silly drawing to illustrate what would be a dream program , essentially each color has a denoted charge amount(-20,15, 4 ,-2 etc. or 0) and then you could see how the particles interact as time progresses.   No idea if this is possible within the framework of your program or if this is a huge undertaking but figured I would put it out there since the program seems to be pretty powerful already.  Thanks for all of this already!!!

**Comment by Syrocco** [comment]
![image](https://github.com/Syrocco/Graphical-EDMD/assets/37348774/e6ff9267-5e43-4cb0-a236-8b940f5a4e79)

I added neutral particles (see (before)last commit), grey-blue particles are neutral and free fly between the others. By hard I meant that to be done "nicely" it requires a bit of work. These types of simulations (event driven) really really like that all particles follow the same rules. Here, I made a particle neutral by simply asking for it to ignore a large chunk of code (or rule), but a lot of computations are still lost for "nothing"...

Yes, the drawing you show is more or less possible to simulate up to some issues:
- The algorithm really dislikes varying sizes (as you might have seen with the button "polydispersity"), especially if we add a potential on top.
- For now we cannot really add particles by "hand" and for very specific initial configurations such as yours we would need to "program" the initial condition by hand or give them through a initial config file (this is partially implemented).
- I'm also using this program for some unrelated work and I don't want to bloat it too much (especially not things that could slow down the main algorithm, but surely I could make an other version!)


Other than that, this should be pretty straightforward to do. Instead of relying on a global potential energy U, we could rely, for each interaction on a particle-particle one. That would for example be U_12 = charge particle 1 * charge particle 2 * mass particle 1 * mass particle 2 with a radius of interaction that depends on the radius of the particle for example (The interactions here, are square well and square shoulders, if you want to look it up, it might not be what you have in mind when you talk about an usual attractive or repulsive potential). If you have in mind a simple initial condition (that can be simply algorithmically described), I can give it a shot!

But for these kind of things, you would be better with LAMMPS (not "event driven" and slower (in single processor), but way more powerful and versatile) or DYNAMO if you want to stay close to the "event driven molecular dynamics" paradigm (that I think Bio people like a lot for proteins and stuff like that). Both of them give you total control over the program through a scripting language (well the scripting language of LAMMPS is a total absurdity, but it is the price to pay for such versatility). You don't really have a real time interface with these programs but they output files containing the positions of the particles, that you can then read with Ovito for example

_______
Edit: I looked up at the Zeta potential and, according to the picture I saw, you indeed need a huge scale difference between the different particles. My program would really perform badly here (well less than 500 particles should be fine but you completely lost advantage of the blaziingly fast speed of the monodisperse algorithm)! Again, for this kind of things, LAMMPS would be the way to go!!

**Comment by Ceres90** [comment]
I tried updating but im not seeing the neutral particles?  does git fetch work?

**Comment by Syrocco** [comment]
Strange
You don't see these two sliders (% of + particles and % of neutral particles) when clicking on charged?
![image](https://github.com/Syrocco/Graphical-EDMD/assets/37348774/fc1b6d0e-8557-424c-b2fc-bba2830be1d9)

In addition to: 
![image](https://github.com/Syrocco/Graphical-EDMD/assets/37348774/0fe875c2-81b7-4704-a976-452ffdf0ea79)
coloring by "charge"?

The commits seem to have worked well:
![image](https://github.com/Syrocco/Graphical-EDMD/assets/37348774/6b390190-d597-4a9b-b70b-574bd77bf388)

**Comment by Ceres90** [comment]
<img width="539" alt="image" src="https://github.com/Syrocco/Graphical-EDMD/assets/95306501/c66a589a-2dfa-4940-8d46-4b51ba0e5881">

**Comment by Ceres90** [comment]
i see color by charge , whats the right way to update the program?  im using "git fetch", "git pull origin main" and it doesn't give me an error but something isn't working

**Comment by Ceres90** [comment]
<img width="389" alt="image" src="https://github.com/Syrocco/Graphical-EDMD/assets/95306501/fe762a84-8661-4fbf-88f2-f676ef150efb">

**Comment by Syrocco** [comment]
Ah that's very strange...  you can try to do git stash and then pull again. Or create a new directory and do a new git clone in it. To start fresh?

**Comment by Ceres90** [comment]
oh didn't realize I had to run "make" after updating!

**Comment by Ceres90** [comment]
its working now

**Comment by Syrocco** [comment]
Cool

**Comment by Ceres90** [comment]
To your earlier comments I may eventually try LAAMP but the learning curve is prohibitively high the fact that I can already play around with your program and that I can watch the results live is pretty nice, even if the results aren't going to be as accurate as LAAMP .   If you are still willing to make changes (whenever you have the time to) I would probably still try this instead.
